### PR TITLE
Collect some small speedups

### PR DESCRIPTION
**Pull request by CyberiaResurrection**
This PR cleans up some slight deduplication in Star.hex_distance / Star.heuristicDistance, and takes the opportunity to memoise ally lookup, and likewise memoising the deduplicated distance core.
On top of that, a couple of other small gains are collected by re-arranging the Star __eq__ check and pre-initialising all fields in Star's constructor

For reference, times reported are under the PyCharm 2020 built-in profiler on the same 4x4.1 Ghz Intel i7-6700k that I did PR https://github.com/makhidkarun/traveller_pyroute/pull/3 on.

On the machine above, I was able to reduce route-gen time:

- at max jump of J-2;
- for Reft and its orthogonally-adjacent sectors (Trojan Reach, Deneb, Gushemege, Verge), Riftspan Reaches and its orthogonally-adjacent sectors (Touchstone and Hlakoi - Trojan Reach and Verge were included earlier), plus Corridor and Vland (to give a giant connected component, as a proxy for bigger runs), 10 sectors total;

from 563 seconds (with no changes) to 521 seconds (with both memoisations and the deduplication) - about 7.5%.  The Star-related changes bring runtime down to 507 seconds - just under 10% reduction in runtime.

### CODE REVIEW

**Review by tjoneslo — CHANGES_REQUESTED** [review]
Please add a note to the first section of the README.md file that the minimum supported version is now Python 3.9+ (or PyPy 7.3 for Python 3.9). I like the idea of using functools.cache, and there may be other places this could be applied. But we should warn the two other users about the version limitations.

**Comment by CyberiaResurrection** [comment]
@tjoneslo , does that last commit do the trick?

File: readme.md [hunk 1]
```diff
@@ -2,7 +2,7 @@ PyRoute for Traveller
 =====================
 
 PyRoute is a trade route generation program for Traveller. PyRoute has been ported to python3, there is a branch with 
-the python2 version.
+the python2 version.  The python3 branch needs Python 3.9+ (or PyPy 7.3).
```
Comment (tjoneslo) [inline readme.md:5]:
```suggestion
the python2 version (no longer maintained).  The released version requires Python 3.9+ (or PyPy 7.3).
```

**Review by tjoneslo — APPROVED** [review]
Made one suggestion for clarifying the requirements. Otherwise LGTM.

**Comment by CyberiaResurrection** [comment]
@tjoneslo , I've rolled in your feedback and rebased the branch atop the last merge.

File: readme.md (hunk 1 above)
Comment (CyberiaResurrection) [reply readme.md]: I think that last commit updates the readme as you suggested, @tjoneslo

**Pull request merged**